boards: support for ST Nucleo L452RE-P

crespum opened the pull request:
This is my first PR so please be kind :)

I wanted to add support for a new development board, the ST Nucleo L452RE-P. There's already support for the L452RE (note the missing P), so I've pretty much copied everything and modified the things that differed:
* The User LED is in PB13
* The MCU supports SMPS. It's not being used by Zephyr at the moment, but I made sure to add a note in the docs highlighting this difference.

In the process of adding this new board I've detected some minor errors in the L452RE documentation which I've fixed.

Finally, according to the user manual, the Zephyr console should be connected to LPUART1 instead of UART2. The pins are the same and I've made sure it works, however I haven't modify this neither in the mux nor in the DT because the corresponding alternate function is not defined in `pinmux_stm32l4x.h`.

ABOSTM reviewed (changes requested):
Hi @crespum thanks for this PullRequest.
And congratulations as it is your 1st one.
I am fine with your 1st commit concerning the issue found in nucleo_l452re.
About the 2nd commit, knowing that 'nucleo_l452re_p' is very close to 'nucleo_l452re',
I would propose a different approach to factorize sources as much as possible (and to reduce further maintenance):
Put both boards in the same directory (nucleo_l452re).
You can take example with directory "zephyr\boards\arm\stm32_min_dev\" which includes 2 boards 
'stm32_min_dev_black' and 'stm32_min_dev_blue'.
As you can see there is a *.dts file for each board which allow you to put the difference (led gpio) and you can include a create a common dts for the common part .
As you can see there is also a *_defconfig file for each board which allows (if necessary in the future) to take into account SMPS difference.

That said, you should create 2 different *.dtsi file for arduino_r3_connector, because there are some differences in gpio routing to arduino connector (see User Manual).

Finally, I agree that User manual is ambiguous about usage of  ST-link Virtual COM port: it points out that user have the choice between LPUART1 (default) and USART1 (by Solder brige hardware changes).
And this is valid whatever soc solder on this board MB1319. But it hide the fact that, at least on this nucleo_l452re_p board,  USART2 is also on the same pins can can be used too.

martinjaeger commented on boards/arm/nucleo_l452re/doc/index.rst:
```diff
@@ -150,10 +150,15 @@ Default Zephyr Peripheral Mapping:
 - UART_1_RX : PA10
 - UART_2_TX : PA2
-- UART_2_RX : PA15
+- UART_2_RX : PA3
```
This should also be updated in the `pinmux.c`.

crespum replied:
It's already correct in `pinmux.c`, I guess it was just a typo in the docs!

martinjaeger reviewed (changes requested):
Congrats to the first PR.

Fun fact: The nucleo_l452re board was actually my first PR a few months ago. And I absolutely can't remember anymore why I took `PA15` for `UART_2_RX`. Your fix makes much more sense.

crespum commented:
Thanks for the feedback, I like the idea of putting the two boards inside the same folder. I'll get to it ASAP.

martinjaeger commented on boards/arm/nucleo_l452re/doc/index.rst:
```diff
@@ -6,11 +6,17 @@ ST Nucleo L452RE
+- ST Nucleo L452RE
+- ST Nucleo L452RE-P
+
+On one hand, the Nucleo L452RE board features an ARM Cortex-M4 based STM32L452RE MCU.
+On the other hand, the Nucleo L452RE-P board features an ARM Cortex-M4 based
+STM32L452RET6PU MCU. Here some highlights of these boards:
```
I would not go into detail regarding the package of the -P version here. Both MCUs are STM32L452RE.

```suggestion
Here some highlights of the boards:
```

martinjaeger commented on boards/arm/nucleo_l452re/nucleo_l452re_p.dts:
```diff
@@ -0,0 +1,93 @@
+#include "arduino_r3_connector.dtsi"
+
+/ {
+	model = "STMicroelectronics STM32L452RE-NUCLEO board";
+	compatible = "st,stm32l452re-nucleo", "st,stm32l452";
```
add `-P`

martinjaeger reviewed (changes requested):
The failing build should be fixed. Otherwise just some nits comments.

ABOSTM commented on boards/arm/nucleo_l452re/doc/index.rst:
```diff
@@ -21,13 +27,18 @@ some highlights of the Nucleo L452RE board:
-  :width: 250px
+.. image:: img/nucleo_l452re_p.jpg
+  :width: 363px
   :align: center
-  :height: 188px
-  :alt: Nucleo L452RE
+  :height: 408px
+  :alt: Nucleo L452RE-P
```
Why replacing nucleo_l452re.jpg by nucleo_l452re_p.jpg ?
This directory includes both boards, so we should have both.
And also restore the binary itself (nucleo_l452re_p.jpg)

crespum replied:
I've just kept `nucleo_l452re_p.jpg` because the quality is far better (see [the current one](https://docs.zephyrproject.org/latest/_images/nucleo_l452re.jpg)). I don't think there's a need to keep both pictures in the repo because you won't notice the differences at first sight.

martinjaeger replied:
I'm actually not a fan of having binary blobs (like images) inside a repo at all, as they bloat the repository size over time. A `git clone` of the Linux kernel is about 3 GB after almost 30 years of development. Zephyr is already at 472 MB after only 4 years. And this doesn't even include the modules and vendor HALs which are out of tree.

In that sense I would vote for just keeping the old image :)

crespum replied:
That I can understand. Please let me know what you prefer.

martinjaeger replied:
I prefer reusing the existing one, but it's not a blocker for me, that's why I approved already.

ABOSTM reviewed (changes requested):
Going 1 step further :
put all common part of nucleo_l452re(_p).dts in a common file (for example nucleo_l452re_common.dtsi)
Keep only specific part (led) in nucleo_l452re(_p).dts, and just include the common file.
You can take 'stm32_min_dev' as example

crespum commented:
Please @ABOSTM check if the DTs are okay now (good idea btw).

ABOSTM commented on boards/arm/nucleo_l452re/nucleo_l452re_p.dts:
```diff
@@ -0,0 +1,28 @@
+#include "nucleo_l452re_common.dtsi"
+
+/ {
+	model = "STMicroelectronics STM32L452RExxPx-NUCLEO board";
```
No needs for xx
```suggestion
	model = "STMicroelectronics STM32L452RE-P-NUCLEO board";
```

ABOSTM reviewed (changes requested):
dts is good. Just a remark about board name.

martinjaeger commented on boards/arm/nucleo_l452re/doc/index.rst:
```diff
@@ -136,24 +145,36 @@ input/output, pull-up, etc.
    :alt: Nucleo L452RE Pinout
 
-For mode details please refer to `STM32 Nucleo-32 board User Manual`_.
+.. image:: img/nucleo_l452re_pinout_p.png
```
Must be `img/nucleo_l452re_p_pinout.png` (see failing CI warning).

martinjaeger commented on boards/arm/nucleo_l452re/doc/index.rst:
```diff
@@ -136,24 +145,36 @@ input/output, pull-up, etc.
+   :height: 446px
+   :alt: Nucleo L452RE-P Pinout
+
+For mode details please refer to `ST Nucleo L452RE User Manual`_ or
```
Just found this typo:
```suggestion
For more details please refer to `ST Nucleo L452RE User Manual`_ or
```

martinjaeger reviewed (changes requested):
Please squash the 3 last commits adding the new board into a single one and add at least one additional line for explanation to the commit message (in addition to the sign-off) to satisfy the CI.

After that we should be ready to go.

crespum committed:
boards: arm: l452re: fix docs

* fixed package version included in the devboard
* fixed uart2_rx pin/port
* added missing peripheral pinouts
* modify image name to follow the naming of the rest of the boards

Signed-off-by: Xabier Crespo Álvarez <x.crespo@scrobotics.es>

crespum committed:
boards: arm: l452re_p: initial support for dev board

Support for the Nucleo L452RE-P has been added. Since it's very similar
to L452RE board, both of them have been added under the same directory.
The main difference is that the P version contains a SMPS but Zephyr is
not taking advantage of it yet.

I've tested the results in a real L452RE-P board.

Signed-off-by: Xabier Crespo Álvarez <x.crespo@scrobotics.es>

martinjaeger commented:
See [here](https://docs.zephyrproject.org/latest/contribute/index.html#commit-message-body) for explanation how to fix the failing Compliance check (change summary message body missing).

crespum commented:
I've just pushed the fixes, however before closing the PR I ran a test and I'm getting the following error:

```
➜ west build -b nucleo_l452re_p -- -G'Unix Makefiles'
-- west build: generating a build system
Including boilerplate (Zephyr base): /home/xxxxxx/zephyr/cmake/app/boilerplate.cmake
-- Application: /home/xxxxxx/app
-- Zephyr version: 2.3.99 (/home/xxxxxx/zephyr)
-- Found Python3: /usr/bin/python3.8 (found suitable exact version "3.8.3") found components: Interpreter 
-- Board: nucleo_l452re_p
-- Found toolchain: zephyr (/home/xxxxxx/zephyr-0.11.2)
-- Found west: /home/xxxxxx/.local/bin/west (found suitable version "0.7.2", minimum required is "0.7.1")
-- Found dtc: /home/xxxxxx/zephyr-0.11.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /home/xxxxxx/zephyr/boards/arm/nucleo_l452re/nucleo_l452re_p.dts
Error: nucleo_l452re_p.dts.pre.tmp:510.1-9 syntax error
FATAL ERROR: Unable to parse input tree
CMake Error at /home/xxxxxx/zephyr/cmake/dts.cmake:202 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /home/xxxxxx/zephyr/cmake/app/boilerplate.cmake:510 (include)
  /home/xxxxxx/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:24 (include)
  /home/xxxxxx/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:35 (include_boilerplate)
  CMakeLists.txt:4 (find_package)


-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/home/xxxxxx/app/build -S/home/xxxxxx/app -GNinja -DBOARD=nucleo_l452re_p '-GUnix Makefiles'
```

Any clues of what's going on?

martinjaeger commented:
> Error: nucleo_l452re_p.dts.pre.tmp:510.1-9 syntax error

Above line gives you a hint where to look for more details. The file is located under `zephhyr/build` and something in line 510 is wrong.

crespum commented:
Alright, it turns out I had to remove `#include <st/l4/stm32l452Xe.dtsi>` from the dts because it's already in the common DT. Thanks for your help!

If the CI passes, the PR should be over unless you want me to modify the image.

ABOSTM reviewed (approved):
LGTM, I have no opinion about images.
I let zephyr maintainers decide, but this may be a decision to take and apply for all boards ...

martinjaeger reviewed (approved):
Great, thanks for the final update. I'm happy with the higher quality image.

carlescufi commented:
Thanks for this first PR @crespum!
Thanks @ABOSTM and @martinjaeger for the reviews.
